cmd/maya-exporter/app/collector/cstorcollector.go

```diff
@@ -0,0 +1,336 @@
+// duplicate descriptors. Those duplicates are simply ignored. However,
+// two different Collectors must not send duplicate descriptors.) This
+// method idempotently sends the same descriptors throughout the
+// lifetime of the Collector. If a Collector encounters an error while
+// executing this method, it must send an invalid descriptor (created
+// with NewInvalidDesc) to signal the error to the registry.
+
+// Describe describes all the registered stats metrics from the OpenEBS volumes.
+func (c *CstorStatsExporter) Describe(ch chan<- *prometheus.Desc) {
+	for _, gauge := range c.gaugesList() {
+		gauge.Describe(ch)
+	}
+
+	for _, counter := range c.countersList() {
+		counter.Describe(ch)
+	}
+}
+
+// Collect is called by the Prometheus registry when collecting
+// metrics. The implementation sends each collected metric via the
+// provided channel and returns once the last metric has been sent. The
+// descriptor of each sent metric is one of those returned by
+// Describe. Returned metrics that share the same descriptor must differ
+// in their variable label values. This method may be called
+// way. Bloc	king occurs at the expense of total performance of rendering
+// concurrently and must therefore be implemented in a concurrency safe
+// all registered metrics. Ideally, Collector implementations support
+// concurrent readers.
+
+// Collect collects all the registered stats metrics from the OpenEBS volumes.
+// It tries to reconnect with the volume if there is any error via a goroutine.
+func (c *CstorStatsExporter) Collect(ch chan<- prometheus.Metric) {
+	// no need to catch the error as exporter should work even if
+	// there are failures in collecting the metrics due to connection
+	// issues or anything else.
+	_ = c.collector()
+	// collect the metrics extracted by collect method
+	for _, gauge := range c.gaugesList() {
+		gauge.Collect(ch)
+	}
+	for _, counter := range c.countersList() {
+		counter.Collect(ch)
+	}
+}
+
+// collector selects the container attached storage for the collection of
+// metrics.Supported CAS are jiva and cstor.
+func (c *CstorStatsExporter) collector() error {
+
+	if c.Conn == nil {
+		// initiate the connection again if connection with the istgt closed
+		// due to timeout or some other errors from istgt side.
+		c.Metrics.connectionRetryCounter.WithLabelValues("Connection closed from cstor, retry").Inc()
+		if c.Conn = InitiateConnection(); c.Conn == nil {
+			glog.Error("Error in initiating the connection")
+			return errors.New("error in initiating connection with socket")
+		}
+	} else {
+		// collect the metrics from cstor if cstor is reachable, else set nil to
+		// the value of net.Conn in case of error so that new connection be created
+		// after the new request from prometheus comes.
+		if err := c.collect(); err != nil {
+			c.Metrics.connectionErrorCounter.WithLabelValues(err.Error()).Inc()
+			glog.Error("Error in connection, closing the connection")
+			c.Conn = nil
```

Comment:
Can we call Conn.close() before making it null ? Again, need to be careful that we should be closing an already closed connection.

Reply:
roger that.